File: cpp/src/analytics/factors.cpp
/**
 * AlgoForge — src/analytics/factors.cpp
 * Fama-French-style multi-factor OLS regression.
 *
 * Builds design matrix X = [1, f1, f2, ...] and calls the OLS solver.
 */

#include "core/analytics.hpp"

#include <cmath>
#include <span>
#include <vector>

namespace algoforge::analytics {

FactorResult factor_regression(
    std::span<const double>                                       strategy_returns,
    const std::unordered_map<std::string, std::vector<double>>&   factor_returns
) noexcept {
    FactorResult fr;

    const int n = static_cast<int>(strategy_returns.size());
    if (n < 2 || factor_returns.empty()) return fr;

    /* ── Collect factor names in sorted order for reproducibility ── */
    std::vector<std::string> names;
    names.reserve(factor_returns.size());
    for (const auto& [name, _] : factor_returns) names.push_back(name);
    std::sort(names.begin(), names.end());

    /* ── Verify all factor series have at least n values ── */
    for (const auto& nm : names) {
        if (static_cast<int>(factor_returns.at(nm).size()) < n) {
            /* Truncate n to the shortest factor series */
        }
    }
    /* Find actual usable length */
    int usable = n;
    for (const auto& nm : names) {
        int sz = static_cast<int>(factor_returns.at(nm).size());
        if (sz < usable) usable = sz;
    }
    if (usable < 2) return fr;

    const int k = static_cast<int>(names.size());

    /* ── Build design matrix X: n x (k+1) with intercept column ── */
    std::vector<std::vector<double>> X(usable, std::vector<double>(k + 1, 0.0));
    for (int i = 0; i < usable; ++i) {
        X[i][0] = 1.0;   /* intercept */
        for (int j = 0; j < k; ++j) {
            X[i][j + 1] = factor_returns.at(names[j])[i];
        }
    }

    /* ── Run OLS ── */
    std::span<const double> y_span(strategy_returns.data(), usable);
    OLSResult ols = ols_regression(X, y_span);

    if (ols.beta.empty()) return fr;

    fr.alpha        = ols.beta[0];
    fr.factor_names = names;
    fr.betas.assign(ols.beta.begin() + 1, ols.beta.end());
    fr.t_stats.assign(ols.t_stats.begin() + 1, ols.t_stats.end());
    fr.r2           = ols.r2;
    fr.adj_r2       = ols.adj_r2;

    return fr;
}

} /* namespace algoforge::analytics */

File: js/src/analytics/factors.js
/**
 * factors.js — Fama-French-style multi-factor OLS regression.
 * Spec §4 (Factors section).
 * Uses the same OLS solver from ml_attribution.js.
 */

import { ols } from './ml_attribution.js';

/**
 * Multi-factor regression: strategy_returns = alpha + sum(beta_f * factor_f) + eps
 *
 * @param {number[]}              strategyReturns  n bar returns of the strategy
 * @param {Record<string,number[]>} factorReturns   dict of factor name -> n bar returns
 * @returns {{
 *   alpha:       number,
 *   betas:       Record<string, number>,
 *   tStats:      Record<string, number>,
 *   stdErrors:   Record<string, number>,
 *   r2:          number,
 *   adjR2:       number,
 *   residuals:   number[],
 * }|null}
 */
export function factorRegression(strategyReturns, factorReturns) {
    const factorNames = Object.keys(factorReturns);
    const n = strategyReturns.length;
    if (n === 0) return null;

    // Build design matrix: [1, f1, f2, ...]
    const X = [];
    for (let i = 0; i < n; i++) {
        const row = [1];
        for (const name of factorNames) {
            row.push(factorReturns[name][i] ?? 0);
        }
        X.push(row);
    }

    const result = ols(X, strategyReturns);
    if (result === null) return null;

    const { beta, residuals, r2, adjR2, tStats, stdErrors } = result;

    // beta[0] = alpha (intercept), beta[1..k] = factor betas
    const betas    = {};
    const tStatMap = {};
    const stdErrMap = {};
    for (let f = 0; f < factorNames.length; f++) {
        const name = factorNames[f];
        betas[name]    = beta[f + 1];
        tStatMap[name] = tStats[f + 1];
        stdErrMap[name] = stdErrors[f + 1];
    }

    return {
        alpha:      beta[0],
        betas,
        tStats:     tStatMap,
        stdErrors:  stdErrMap,
        r2,
        adjR2,
        residuals,
    };
}
